File: app/src/main/java/com/example/paetz/yacguide/database/AppDatabase.java
package com.example.paetz.yacguide.database;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.arch.persistence.room.TypeConverters;
import android.content.Context;

import com.example.paetz.yacguide.database.Comment.RegionComment;
import com.example.paetz.yacguide.database.Comment.RegionCommentDao;
import com.example.paetz.yacguide.database.Comment.RockComment;
import com.example.paetz.yacguide.database.Comment.RockCommentDao;
import com.example.paetz.yacguide.database.Comment.RouteComment;
import com.example.paetz.yacguide.database.Comment.RouteCommentDao;
import com.example.paetz.yacguide.database.Comment.SectorComment;
import com.example.paetz.yacguide.database.Comment.SectorCommentDao;
import com.example.paetz.yacguide.utils.Converters;

@Database(entities = {Country.class, Region.class, Sector.class, Rock.class, Route.class, Ascend.class, Partner.class,
        RegionComment.class, SectorComment.class, RockComment.class, RouteComment.class}, version = 1)
@TypeConverters({Converters.class})
public abstract class AppDatabase extends RoomDatabase {

    private static AppDatabase _instance;

    public static final String UNKNOWN_NAME = "???";
    public static final int INVALID_ID = -1;

    public abstract CountryDao countryDao();
    public abstract RegionDao regionDao();
    public abstract SectorDao sectorDao();
    public abstract RockDao rockDao();
    public abstract RouteDao routeDao();
    public abstract AscendDao ascendDao();
    public abstract PartnerDao partnerDao();
    public abstract RegionCommentDao regionCommentDao();
    public abstract SectorCommentDao sectorCommentDao();
    public abstract RockCommentDao rockCommentDao();
    public abstract RouteCommentDao routeCommentDao();

    public static AppDatabase getAppDatabase(Context context) {
        if (_instance == null) {
            _instance = Room.databaseBuilder(context.getApplicationContext(),
                    AppDatabase.class, "yac_database").allowMainThreadQueries().build();
        }
        return _instance;
    }

    public static void destroyInstance() {
        _instance = null;
    }

    // Helpers to avoid code duplications
    public void deleteCountries() {
        final Country[] countries = countryDao().getAll();
        for (final Country country : countries) {
            deleteRegions(country.getName());
        }
        countryDao().deleteAll();
    }

    public void deleteRegions(String countryName) {
        final Region[] regions = regionDao().getAll(countryName);
        for (final Region region : regions) {
            deleteSectors(region.getId());
        }
        regionDao().deleteAll(countryName);
    }

    public void deleteSectors(int regionId) {
        final Sector[] sectors = sectorDao().getAll(regionId);
        for (final Sector sector : sectors) {
            deleteRocks(sector.getId());
        }
        sectorDao().deleteAll(regionId);
        regionCommentDao().deleteAll(regionId);
    }

    public void deleteRocks(int sectorId) {
        final Rock[] rocks = rockDao().getAll(sectorId);
        for (final Rock rock : rocks) {
            final Route[] routes = routeDao().getAll(rock.getId());
            for (final Route route : routes) {
                routeCommentDao().deleteAll(route.getId());
            }
            routeDao().deleteAll(rock.getId());
            rockCommentDao().deleteAll(rock.getId());
        }
        rockDao().deleteAll(sectorId);
        sectorCommentDao().deleteAll(sectorId);
    }

    // Some default stuff necessary for tourbook if according objects have
    // been deleted from the database
    public Route createUnknownRoute() {
        Route route = new Route();
        route.setName(UNKNOWN_NAME);
        route.setGrade(UNKNOWN_NAME);
        return route;
    }
    public Rock createUnknownRock() {
        Rock rock = new Rock();
        rock.setName(UNKNOWN_NAME);
        return rock;
    }
    public Sector createUnknownSector() {
        Sector sector = new Sector();
        sector.setName(UNKNOWN_NAME);
        return sector;
    }
    public Region createUnknownRegion() {
        Region region = new Region();
        region.setName(UNKNOWN_NAME);
        return region;
    }
}

File: app/src/main/java/com/yacgroup/yacguide/database/AppDatabase.kt
/*
 * Copyright (C) 2019 Fabian Kantereit
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.yacgroup.yacguide.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters

import com.yacgroup.yacguide.database.Comment.RegionComment
import com.yacgroup.yacguide.database.Comment.RegionCommentDao
import com.yacgroup.yacguide.database.Comment.RockComment
import com.yacgroup.yacguide.database.Comment.RockCommentDao
import com.yacgroup.yacguide.database.Comment.RouteComment
import com.yacgroup.yacguide.database.Comment.RouteCommentDao
import com.yacgroup.yacguide.database.Comment.SectorComment
import com.yacgroup.yacguide.database.Comment.SectorCommentDao
import com.yacgroup.yacguide.utils.Converters

/*
 * Note: Please do not use this class directly for accessing the database.
 *       DatabaseWrapper provides all necessary methods with much better runtime performance.
 */

@Database(entities = [Country::class, Region::class, Sector::class, Rock::class, Route::class, Ascend::class, Partner::class, RegionComment::class, SectorComment::class, RockComment::class, RouteComment::class], version = 1)
@TypeConverters(Converters::class)
abstract class AppDatabase : RoomDatabase() {

    abstract fun countryDao(): CountryDao
    abstract fun regionDao(): RegionDao
    abstract fun sectorDao(): SectorDao
    abstract fun rockDao(): RockDao
    abstract fun routeDao(): RouteDao
    abstract fun ascendDao(): AscendDao
    abstract fun partnerDao(): PartnerDao
    abstract fun regionCommentDao(): RegionCommentDao
    abstract fun sectorCommentDao(): SectorCommentDao
    abstract fun rockCommentDao(): RockCommentDao
    abstract fun routeCommentDao(): RouteCommentDao

    companion object {

        private var _instance: AppDatabase? = null

        fun getAppDatabase(context: Context): AppDatabase {
            synchronized(AppDatabase::class) {
                if (_instance == null) {
                    _instance = Room.databaseBuilder(context.applicationContext,
                            AppDatabase::class.java, "yac_database").allowMainThreadQueries().build()
                }
                return _instance as AppDatabase
            }
        }
    }
}
